src/myrtlespeech/data/dataset/commonvoice.py

```diff
@@ -0,0 +1,307 @@
+
+    def _check_subset_integrity(self, subset: str) -> bool:
+        path = os.path.join(self.root, self.base_dir, subset + ".tsv")
+        try:
+            actual_md5 = utils.checksum_file(path, "md5")
+        except (FileNotFoundError, NotADirectoryError, PermissionError):
+            return False
+        return actual_md5 == self.subset_hashes[subset]
+
+    def check_integrity(self) -> None:
+        """Returns True if each subset in ``self.subsets`` is valid."""
+        for subset in self.subsets:
+            if not self._check_subset_integrity(subset):
+                raise ValueError(f"subset {subset} not found or corrupt")
+
+        # Checks clips directory
+        path = os.path.join(self.root, self.base_dir, "clips")
+        try:
+            actual_md5 = utils.checksum_dir(path, "md5")
+        except (FileNotFoundError, NotADirectoryError, PermissionError):
+            raise ValueError(f"{self.base_dir} directory not found")
+        if actual_md5 != self.clips_hash:
+            raise ValueError(f"{self.base_dir} directory corrupt")
+
+    def load_data(self) -> None:
+        """Loads the data from disk."""
+        self.paths: List[str] = []
+        self.durations: List[float] = []
+        self.transcriptions: List[str] = []
+
+        os.makedirs(
+            os.path.join(self.root, self.base_dir, "clips_down"), exist_ok=True
+        )
+        for subset in self.subsets:
+            self._parse_subset_file(
+                os.path.join(self.root, self.base_dir), subset
+            )
+
+        self._sort_by_duration()
+
+    def _parse_subset_file(self, root: str, name: str) -> None:
+        """Parses each sample in a transcription file."""
+        trans_path = os.path.join(root, name + ".tsv")
+        with open(trans_path, "r", encoding="utf-8") as trans:
+            trans.readline()  # First line is header
+            # Each line has the form "ID \t PATH \t TRANSCRIPT \t <METADATA>"
+            for line in trans:
+                parts = line.split("\t")
+                path = parts[1]
+                transcript = parts[2]
+                dropped = self._process_audio(root, path)
+                if not dropped:
+                    self._process_transcript(transcript)
+
+    def _process_audio(self, root: str, file: str) -> bool:
+        """Returns True if sample was dropped due to being too long."""
+        path = os.path.join(root, "clips", file)
+        si, _ = torchaudio.info(path)
+        duration = (si.length / si.channels) / si.rate
+        if self.max_duration is not None and duration > self.max_duration:
+            return True
+
+        down_file = file[:-3] + "wav"
+        down_path = os.path.join(root, "clips_downsampled", down_file)
+        if not os.path.isfile(down_path):
+            # Convert mp3 -> wav and downsample to 16KHz
+            audio, rate = torchaudio.load(path)
+            torchaudio.save(down_path, audio, 16000)
```

Comment:
Does the resampling need to be done offline now that `torchaudio` is being used? 

If it's possible to resample when calling `__getitem__` then this may be preferable for now. Note that currently each time the dataset is created `load_data` is called causing the entire dataset to be loaded, downsampled, and saved.

Reply:
It looks like the downsampling is only done once. If the downsampled file already exists then `os.path.isfile(down_path)` will return true and so the downsampling and saving of the new file will be skipped. It will still load in every file to check its length but LS does this too.

Reply:
Woops, good spot, missed that check.